lib/cache/cache.go

```diff
@@ -517,18 +722,28 @@ func (c *Cache) notify(event Event) {
 //   we assume that this cache will eventually end up in a correct state
 //   potentially lagging behind the state of the database.
 //
-func (c *Cache) fetchAndWatch(ctx context.Context, retry utils.Retry) error {
+func (c *Cache) fetchAndWatch(ctx context.Context, retry utils.Retry, timer *time.Timer) error {
 	watcher, err := c.Events.NewWatcher(c.ctx, services.Watch{
 		QueueSize:       c.QueueSize,
 		Name:            c.Component,
 		Kinds:           c.watchKinds(),
 		MetricComponent: c.MetricComponent,
 	})
 	if err != nil {
-		c.notify(Event{Type: WatcherFailed})
+		c.notify(c.ctx, Event{Type: WatcherFailed})
 		return trace.Wrap(err)
 	}
 	defer watcher.Close()
+
+	// ensure that the timer is stopped and drained
+	timer.Stop()
```

Comment:
Have you considered using the if instead of the select:
```go
if !timer.Stop() {
	<-timer.C
}
```
which is more readable imo.

Reply:
That pattern is for cases where you always know that `timer.C` was not consumed.  We could track whether the previous call to `fetchAndWatch()` resulted in a timeout, which would allow us to be certain whether or not `timer.C` needed to be drained (i.e. `if !timer.Stop() && !lastCallTimedOut { ... }`), but IMO thats a lot of plumbing for minimal gain.  A nonblocking recv does the job just as well, and is less brittle.